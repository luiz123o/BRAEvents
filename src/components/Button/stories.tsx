import { Story, Meta } from '@storybook/react/types-6-0'
import { AddShoppingCart } from '@styled-icons/material-twotone'
import Button from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />

Default.args = {
  children: 'Compre já'
}
export const withIcon: Story = (args) => <Button {...args} />

withIcon.args = {
  size: 'small',
  children: 'Compre já',
  icon: <AddShoppingCart />
}
export const asLink: Story = (args) => <Button {...args} />

asLink.args = {
  size: 'large',
  children: 'Compre já',
  as: 'a',
  href: '/link'
}
