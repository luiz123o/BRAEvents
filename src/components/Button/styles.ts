import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'fullWidth' | 'minimal'>

const wrapperMatchModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  minimal: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};
    &:hover {
      transition: color 0.2s ease-in-out;
      color: ${theme.colors.secondary};
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, hasIcon, minimal }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-decoration: none;
    background: rgb(180, 102, 200);
    background: linear-gradient(
      180deg,
      rgba(180, 102, 200, 1) 0%,
      rgba(136, 105, 224, 1) 93%
    );
    color: ${theme.colors.white};
    cursor: pointer;
    border: none;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};

    &:hover {
      transition: ${minimal ? 'none' : '0.4s'};
      background: ${minimal
        ? 'none'
        : `linear-gradient(180deg, #994cac 0%, #704ed0 100%)`};
    }

    ${!!size && wrapperMatchModifiers[size](theme)};
    ${!!fullWidth && wrapperMatchModifiers.fullWidth()};
    ${!!hasIcon && wrapperMatchModifiers.withIcon(theme)};
    ${!!minimal && wrapperMatchModifiers.minimal(theme)};
  `}
`
