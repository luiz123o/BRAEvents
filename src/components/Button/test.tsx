import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-twotone'
import Button from '.'

describe('<Button />', () => {
  it('should render the medium button default', () => {
    const { container } = renderWithTheme(<Button>Compre já</Button>)

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render the small size', () => {
    renderWithTheme(<Button size="small">Compre já</Button>)

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })
  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Compre já</Button>)

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem',
      padding: '0.8rem 4.8rem'
    })
  })
  it('should render fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Compre já</Button>)

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyle({
      width: '100%'
    })
  })
  it('should render an icon version', () => {
    renderWithTheme(
      <Button fullWidth icon={<AddShoppingCart data-testid="icon" />}>
        Compre já
      </Button>
    )

    expect(screen.getByText(/Compre já/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Compre já
      </Button>
    )

    expect(screen.getByRole('link', { name: /Compre já/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
  it('should render a minimal version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />} minimal>
        Compre já
      </Button>
    )

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyle({
      background: 'none',
      color: '#7B95DE'
    })

    expect(screen.getByRole('button', { name: /Compre já/i })).toHaveStyleRule(
      'background',
      'none',
      {
        modifier: ':hover'
      }
    )
  })
})
