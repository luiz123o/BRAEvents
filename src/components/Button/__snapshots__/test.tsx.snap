// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium button default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: rgb(180,102,200);
  background: linear-gradient( 180deg,rgba(180,102,200,1) 0%,rgba(136,105,224,1) 93% );
  color: #fafafa;
  cursor: pointer;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  -webkit-transition: 0.4s;
  transition: 0.4s;
  background: linear-gradient(180deg,#994cac 0%,#704ed0 100%);
}

<button
  class="c0"
>
  <span>
    Compre j√°
  </span>
</button>
`
